### ISSUE TITLE
MSN Loss computation 

### ISSUE THREAD

**Issue by SaeedShurrab** [lightly-ai/lightly#1439]
Greetings Lightly team,

I am currently working with the MSN model but I noticed something weird in the implementation.

in the example you presented, you concatenated both anchor_out and anchor_focal_out as follows:
`anchors_out = torch.cat([anchors_out, anchors_focal_out], dim=0)`
 
Assume we have 3 images each with 10 focal views. The resulting tensor after concatenation is of dimensionality [33,embedding_dim] and the first 3 rows will be the classification tokens of the 3 random views followed by the 10 classification tokens for the focal views of the first random view, then the second 10 classification of the second random view, and so on.  That is, the arrangement of the tokens will be as follows:
1,2,3,1,1,1,...,2,2,2...,3,3,3

On the other side, the target branch output will be [3, embedding_dim]. In the loss function, you repeat this output 11 times to match the dimensionality of the anchors as follows:
`target_probs = target_probs.repeat((num_views, 1))`
This will repeat the target output but will not match correctly between the target and anchor outputs. That is:

target  vs anchor
1          ->      1
2          ->      2
3          ->      3
1          ->      1
2          ->      1
3          ->      1
            .
            .
            .
1          ->      2
2          ->      2
3          ->      2
            .
            .
            .
1          ->      3
2          ->      3
3          ->      3
            .
            .
            .

Thus, the loss is not computed correctly due to the mismatching between anchors and targets.

Could you please elaborate on this?

**Comment by guarin** [comment]
Hi! I think there is a bit of confusion how the focal anchors are ordered. They are created like this:

```
anchors_focal = torch.concat(views[2:], dim=0)
```

Note that here we concat multiple views together. Each view contains one tensor for every input image in the order 1,2,3. So the order in `anchors_focal` will be 1,2,3,1,2,3,1,2,3.

If you look at the code for `MSNTransform` you can indeed see that we create one transform for every focal view individually:
https://github.com/lightly-ai/lightly/blob/54a1b7bced3e0b34b1f4c47c13f5eff2347bcbb4/lightly/transforms/msn_transform.py#L131-L132

This is why we get many focal views with 3 images each.

Hope this helps, and please let me know if I missed something :)

**Comment by guarin** [comment]
Closing this for now, please reopen if you have more questions.
